File: src/engine/math3d.py
"""
math3d.py -- 轻量 3D 数学库

《远行假设 · 谁》自研引擎的数学基础层。
全部基于 numpy float32，矩阵采用列主序（OpenGL 约定），
即 M @ v 形式做变换，上传 GL 时使用 .T.tobytes() 或 bytes(M.T)。

约定:
  - 右手坐标系, +Y 向上, -Z 为相机前方
  - 角度单位统一为弧度, 对外接口若用角度会显式标注 _deg
"""
from __future__ import annotations

import math

import numpy as np

F32 = np.float32


# --------------------------------------------------------------------------
# 向量
# --------------------------------------------------------------------------
def vec3(x=0.0, y=0.0, z=0.0) -> np.ndarray:
    return np.array([x, y, z], dtype=F32)


def vec4(x=0.0, y=0.0, z=0.0, w=1.0) -> np.ndarray:
    return np.array([x, y, z, w], dtype=F32)


def normalize(v: np.ndarray) -> np.ndarray:
    n = float(np.linalg.norm(v))
    if n < 1e-9:
        return np.zeros_like(v, dtype=F32)
    return (v / n).astype(F32)


def length(v: np.ndarray) -> float:
    return float(np.linalg.norm(v))


def length2(v: np.ndarray) -> float:
    return float(np.dot(v, v))


def cross(a: np.ndarray, b: np.ndarray) -> np.ndarray:
    return np.cross(a, b).astype(F32)


def dot(a: np.ndarray, b: np.ndarray) -> float:
    return float(np.dot(a, b))


def lerp(a, b, t):
    return a + (b - a) * t


def clamp(x, lo, hi):
    return lo if x < lo else (hi if x > hi else x)


def smoothstep(e0, e1, x):
    t = clamp((x - e0) / (e1 - e0 + 1e-9), 0.0, 1.0)
    return t * t * (3.0 - 2.0 * t)


def damp(current, target, smoothing, dt):
    """帧率无关的指数插值。smoothing 越小收敛越快。"""
    return lerp(current, target, 1.0 - math.pow(smoothing, dt))


def move_towards(current, target, max_delta):
    d = target - current
    if abs(d) <= max_delta:
        return target
    return current + math.copysign(max_delta, d)


def angle_lerp(a, b, t):
    """角度插值, 处理 -pi..pi 环绕。"""
    d = (b - a + math.pi) % (2 * math.pi) - math.pi
    return a + d * t


# --------------------------------------------------------------------------
# 矩阵
# --------------------------------------------------------------------------
def identity() -> np.ndarray:
    return np.identity(4, dtype=F32)


def translate(x, y, z) -> np.ndarray:
    m = identity()
    m[0, 3] = x
    m[1, 3] = y
    m[2, 3] = z
    return m


def scale(x, y=None, z=None) -> np.ndarray:
    if y is None:
        y = z = x
    m = identity()
    m[0, 0] = x
    m[1, 1] = y
    m[2, 2] = z
    return m


def rotate_x(a) -> np.ndarray:
    c, s = math.cos(a), math.sin(a)
    m = identity()
    m[1, 1], m[1, 2] = c, -s
    m[2, 1], m[2, 2] = s, c
    return m


def rotate_y(a) -> np.ndarray:
    c, s = math.cos(a), math.sin(a)
    m = identity()
    m[0, 0], m[0, 2] = c, s
    m[2, 0], m[2, 2] = -s, c
    return m


def rotate_z(a) -> np.ndarray:
    c, s = math.cos(a), math.sin(a)
    m = identity()
    m[0, 0], m[0, 1] = c, -s
    m[1, 0], m[1, 1] = s, c
    return m


def compose_trs(pos, rot_y=0.0, scl=1.0) -> np.ndarray:
    """最常用的 平移*绕Y旋转*缩放 组合, 避免通用矩阵乘的开销。"""
    c, s = math.cos(rot_y), math.sin(rot_y)
    if np.isscalar(scl):
        sx = sy = sz = float(scl)
    else:
        sx, sy, sz = float(scl[0]), float(scl[1]), float(scl[2])
    m = np.zeros((4, 4), dtype=F32)
    m[0, 0] = c * sx
    m[0, 2] = s * sz
    m[1, 1] = sy
    m[2, 0] = -s * sx
    m[2, 2] = c * sz
    m[0, 3] = pos[0]
    m[1, 3] = pos[1]
    m[2, 3] = pos[2]
    m[3, 3] = 1.0
    return m


def perspective(fovy_deg, aspect, znear, zfar) -> np.ndarray:
    f = 1.0 / math.tan(math.radians(fovy_deg) * 0.5)
    m = np.zeros((4, 4), dtype=F32)
    m[0, 0] = f / max(aspect, 1e-6)
    m[1, 1] = f
    m[2, 2] = (zfar + znear) / (znear - zfar)
    m[2, 3] = (2.0 * zfar * znear) / (znear - zfar)
    m[3, 2] = -1.0
    return m


def ortho(left, right, bottom, top, znear, zfar) -> np.ndarray:
    m = identity()
    m[0, 0] = 2.0 / (right - left)
    m[1, 1] = 2.0 / (top - bottom)
    m[2, 2] = -2.0 / (zfar - znear)
    m[0, 3] = -(right + left) / (right - left)
    m[1, 3] = -(top + bottom) / (top - bottom)
    m[2, 3] = -(zfar + znear) / (zfar - znear)
    return m


def look_at(eye, target, up=None) -> np.ndarray:
    if up is None:
        up = vec3(0, 1, 0)
    eye = np.asarray(eye, dtype=F32)
    target = np.asarray(target, dtype=F32)
    f = normalize(target - eye)
    # 前向与 up 平行时换一个 up, 防止叉乘退化
    if abs(dot(f, normalize(up))) > 0.9995:
        up = vec3(0, 0, 1) if abs(f[1]) > 0.9 else vec3(0, 1, 0)
    s = normalize(cross(f, up))
    u = cross(s, f)
    m = identity()
    m[0, :3] = s
    m[1, :3] = u
    m[2, :3] = -f
    m[0, 3] = -dot(s, eye)
    m[1, 3] = -dot(u, eye)
    m[2, 3] = dot(f, eye)
    return m


def mat_bytes(m: np.ndarray) -> bytes:
    """转成 GL 需要的列主序字节流。"""
    return np.ascontiguousarray(m.T, dtype=F32).tobytes()


def inverse(m: np.ndarray) -> np.ndarray:
    return np.linalg.inv(m).astype(F32)


def transform_point(m: np.ndarray, p) -> np.ndarray:
    v = np.array([p[0], p[1], p[2], 1.0], dtype=F32)
    r = m @ v
    if abs(r[3]) > 1e-9:
        r = r / r[3]
    return r[:3].astype(F32)


def transform_dir(m: np.ndarray, d) -> np.ndarray:
    v = np.array([d[0], d[1], d[2], 0.0], dtype=F32)
    return (m @ v)[:3].astype(F32)


# --------------------------------------------------------------------------
# 视锥剔除
# --------------------------------------------------------------------------
class Frustum:
    """从 view-projection 矩阵提取 6 个裁剪平面, 做包围球剔除。"""

    __slots__ = ("planes",)

    def __init__(self, viewproj: np.ndarray):
        m = viewproj
        rows = []
        # left, right, bottom, top, near, far
        rows.append(m[3] + m[0])
        rows.append(m[3] - m[0])
        rows.append(m[3] + m[1])
        rows.append(m[3] - m[1])
        rows.append(m[3] + m[2])
        rows.append(m[3] - m[2])
        planes = np.array(rows, dtype=F32)
        norms = np.linalg.norm(planes[:, :3], axis=1, keepdims=True)
        norms[norms < 1e-9] = 1.0
        self.planes = (planes / norms).astype(F32)

    def sphere_visible(self, center, radius: float) -> bool:
        p = self.planes
        d = p[:, 0] * center[0] + p[:, 1] * center[1] + p[:, 2] * center[2] + p[:, 3]
        return bool(np.all(d >= -radius))

    def spheres_visible(self, centers: np.ndarray, radius: float) -> np.ndarray:
        """批量剔除, centers 形状 (N,3), 返回 bool 掩码。"""
        p = self.planes
        d = centers @ p[:, :3].T + p[:, 3]
        return np.all(d >= -radius, axis=1)


# --------------------------------------------------------------------------
# 噪声 (地形/植被分布共用)
# --------------------------------------------------------------------------
_PERM_SEED = 1337


def _build_perm(seed: int) -> np.ndarray:
    rng = np.random.default_rng(seed)
    p = np.arange(256, dtype=np.int32)
    rng.shuffle(p)
    return np.concatenate([p, p]).astype(np.int32)


_PERM = _build_perm(_PERM_SEED)


def _fade(t):
    return t * t * t * (t * (t * 6 - 15) + 10)


def perlin2(x: np.ndarray, y: np.ndarray, perm: np.ndarray | None = None) -> np.ndarray:
    """向量化 2D Perlin 噪声, 返回大致 [-1,1]。"""
    if perm is None:
        perm = _PERM
    xi = np.floor(x).astype(np.int32) & 255
    yi = np.floor(y).astype(np.int32) & 255
    xf = x - np.floor(x)
    yf = y - np.floor(y)
    u = _fade(xf)
    v = _fade(yf)

    aa = perm[perm[xi] + yi]
    ab = perm[perm[xi] + yi + 1]
    ba = perm[perm[xi + 1] + yi]
    bb = perm[perm[xi + 1] + yi + 1]

    def grad(h, dx, dy):
        h = h & 7
        gx = np.where(h < 4, dx, dy)
        gy = np.where(h < 4, dy, dx)
        sx = np.where((h & 1) == 0, 1.0, -1.0)
        sy = np.where((h & 2) == 0, 1.0, -1.0)
        return gx * sx + gy * sy

    x1 = lerp(grad(aa, xf, yf), grad(ba, xf - 1, yf), u)
    x2 = lerp(grad(ab, xf, yf - 1), grad(bb, xf - 1, yf - 1), u)
    return lerp(x1, x2, v).astype(F32)


def fbm2(x, y, octaves=5, lacunarity=2.0, gain=0.5, perm=None) -> np.ndarray:
    """分形布朗运动, 地形主体。"""
    total = np.zeros_like(np.asarray(x, dtype=F32))
    amp = 1.0
    freq = 1.0
    norm = 0.0
    for _ in range(octaves):
        total = total + perlin2(x * freq, y * freq, perm) * amp
        norm += amp
        amp *= gain
        freq *= lacunarity
    return (total / max(norm, 1e-6)).astype(F32)


def ridged2(x, y, octaves=5, lacunarity=2.0, gain=0.5, perm=None) -> np.ndarray:
    """脊状噪声, 用于山脊线。"""
    total = np.zeros_like(np.asarray(x, dtype=F32))
    amp = 1.0
    freq = 1.0
    norm = 0.0
    for _ in range(octaves):
        n = 1.0 - np.abs(perlin2(x * freq, y * freq, perm))
        total = total + (n * n) * amp
        norm += amp
        amp *= gain
        freq *= lacunarity
    return (total / max(norm, 1e-6)).astype(F32)

File: src/Farwalk/Engine/Math3D.cs
// Math3D.cs — 向量/矩阵/噪声工具 (C# 移植自 Python 版 math3d)
using System;

namespace Farwalk.Engine
{
    public struct Vec3
    {
        public float X, Y, Z;
        public Vec3(float x, float y, float z) { X = x; Y = y; Z = z; }
        public static Vec3 operator +(Vec3 a, Vec3 b) => new(a.X + b.X, a.Y + b.Y, a.Z + b.Z);
        public static Vec3 operator -(Vec3 a, Vec3 b) => new(a.X - b.X, a.Y - b.Y, a.Z - b.Z);
        public static Vec3 operator *(Vec3 a, float s) => new(a.X * s, a.Y * s, a.Z * s);
        public static Vec3 operator *(float s, Vec3 a) => a * s;
        public static Vec3 operator /(Vec3 a, float s) => new(a.X / s, a.Y / s, a.Z / s);
        public float Length() => MathF.Sqrt(X * X + Y * Y + Z * Z);
        public Vec3 Normalized()
        {
            float l = Length();
            return l < 1e-6f ? this : this / l;
        }
        public static float Dot(Vec3 a, Vec3 b) => a.X * b.X + a.Y * b.Y + a.Z * b.Z;
        public static Vec3 Cross(Vec3 a, Vec3 b) => new(
            a.Y * b.Z - a.Z * b.Y,
            a.Z * b.X - a.X * b.Z,
            a.X * b.Y - a.Y * b.X);
        public static Vec3 Lerp(Vec3 a, Vec3 b, float t) => a + (b - a) * t;
        public override string ToString() => $"({X:F2},{Y:F2},{Z:F2})";
    }

    public static class Math3D
    {
        public const float PI = 3.14159265f;
        public const float TAU = 6.2831853f;

        public static float Clamp(float v, float lo, float hi) => v < lo ? lo : (v > hi ? hi : v);
        public static int ClampI(int v, int lo, int hi) => v < lo ? lo : (v > hi ? hi : v);
        public static float Damp(float cur, float tgt, float k, float dt) =>
            cur + (tgt - cur) * (1f - MathF.Exp(-k * 120f * dt));
        public static float Lerp(float a, float b, float t) => a + (b - a) * t;
        public static float Smoothstep(float e0, float e1, float x)
        {
            float t = Clamp((x - e0) / (e1 - e0), 0f, 1f);
            return t * t * (3f - 2f * t);
        }

        public static float[] Mat4Identity()
        {
            var m = new float[16];
            m[0] = m[5] = m[10] = m[15] = 1f;
            return m;
        }

        // 列主序 mat4 (OpenGL 风格)
        public static float[] Perspective(float fovyDeg, float aspect, float zn, float zf)
        {
            float f = 1f / MathF.Tan(fovyDeg * PI / 360f);
            var m = new float[16];
            m[0] = f / aspect; m[5] = f; m[10] = (zf + zn) / (zn - zf); m[11] = -1f;
            m[14] = (2f * zf * zn) / (zn - zf);
            return m;
        }

        public static float[] Ortho(float l, float r, float b, float t, float zn, float zf)
        {
            var m = new float[16];
            m[0] = 2f / (r - l); m[5] = 2f / (t - b); m[10] = -2f / (zf - zn);
            m[12] = -(r + l) / (r - l); m[13] = -(t + b) / (t - b); m[14] = -(zf + zn) / (zf - zn);
            m[15] = 1f;
            return m;
        }

        public static float[] LookAt(Vec3 eye, Vec3 center, Vec3? upHint = null)
        {
            Vec3 up = upHint ?? new Vec3(0, 1, 0);
            Vec3 f = (center - eye).Normalized();
            Vec3 s = Vec3.Cross(f, up).Normalized();
            Vec3 u = Vec3.Cross(s, f);
            var m = new float[16];
            m[0] = s.X; m[1] = u.X; m[2] = -f.X; m[3] = 0;
            m[4] = s.Y; m[5] = u.Y; m[6] = -f.Y; m[7] = 0;
            m[8] = s.Z; m[9] = u.Z; m[10] = -f.Z; m[11] = 0;
            m[12] = -Vec3.Dot(s, eye); m[13] = -Vec3.Dot(u, eye); m[14] = Vec3.Dot(f, eye); m[15] = 1;
            return m;
        }

        // 矩阵乘法 (列主序)
        public static float[] Mul(float[] a, float[] b)
        {
            var o = new float[16];
            for (int c = 0; c < 4; c++)
                for (int r = 0; r < 4; r++)
                {
                    float s = 0;
                    for (int k = 0; k < 4; k++) s += a[k * 4 + r] * b[c * 4 + k];
                    o[c * 4 + r] = s;
                }
            return o;
        }

        // ---- 值噪声 ----
        public static int[] BuildPerm(int seed)
        {
            var p = new int[512];
            var rnd = new Random(seed);
            var base_ = new int[256];
            for (int i = 0; i < 256; i++) base_[i] = i;
            for (int i = 255; i > 0; i--)
            {
                int j = rnd.Next(i + 1);
                (base_[i], base_[j]) = (base_[j], base_[i]);
            }
            for (int i = 0; i < 512; i++) p[i] = base_[i & 255];
            return p;
        }

        static float Grad(int h, float x, float y, float z)
        {
            h &= 15;
            float u = h < 8 ? x : y;
            float v = h < 4 ? y : (h == 12 || h == 14 ? x : z);
            return ((h & 1) == 0 ? u : -u) + ((h & 2) == 0 ? v : -v);
        }

        public static float Noise3(float x, float y, float z, int[] perm)
        {
            int X = (int)MathF.Floor(x) & 255, Y = (int)MathF.Floor(y) & 255, Z = (int)MathF.Floor(z) & 255;
            x -= MathF.Floor(x); y -= MathF.Floor(y); z -= MathF.Floor(z);
            float u = x * x * (3 - 2 * x), v = y * y * (3 - 2 * y), w = z * z * (3 - 2 * z);
            int A = perm[X] + Y, AA = perm[A] + Z, AB = perm[A + 1] + Z;
            int B = perm[X + 1] + Y, BA = perm[B] + Z, BB = perm[B + 1] + Z;
            return Lerp(
                Lerp(Lerp(Grad(perm[AA], x, y, z), Grad(perm[BA], x - 1, y, z), u),
                     Lerp(Grad(perm[AB], x, y - 1, z), Grad(perm[BB], x - 1, y - 1, z), u), v),
                Lerp(Lerp(Grad(perm[AA + 1], x, y, z - 1), Grad(perm[BA + 1], x - 1, y, z - 1), u),
                     Lerp(Grad(perm[AB + 1], x, y - 1, z - 1), Grad(perm[BB + 1], x - 1, y - 1, z - 1), u), v),
                w);
        }

        public static float Fbm2(float x, float y, int oct, float lac, float gain, int[] perm)
        {
            float total = 0, amp = 1, freq = 1, norm = 0;
            for (int i = 0; i < oct; i++)
            {
                total += Noise3(x * freq, y * freq, 0.5f, perm) * amp;
                norm += amp;
                amp *= gain; freq *= lac;
            }
            return total / norm;
        }

        public static float Ridged2(float x, float y, int oct, float lac, float gain, int[] perm)
        {
            float total = 0, amp = 1, freq = 1, norm = 0;
            for (int i = 0; i < oct; i++)
            {
                float n = 1f - MathF.Abs(Noise3(x * freq, y * freq, 0.3f, perm));
                total += n * n * amp;
                norm += amp; amp *= gain; freq *= lac;
            }
            return total / norm;
        }
    }
}
